exp/controllers/awsmanagedcluster_controller.go

```diff
@@ -0,0 +1,192 @@
+	err := r.Get(ctx, req.NamespacedName, awsManagedCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, awsManagedCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	if util.IsPaused(cluster, awsManagedCluster) {
+		log.Info("AWSManagedCluster or linked Cluster is marked as paused. Won't reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	controlPlane := &infrav1exp.AWSManagedControlPlane{}
+	controlPlaneRef := types.NamespacedName{
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+	}
+
+	if err := r.Get(ctx, controlPlaneRef, controlPlane); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get control plane ref: %w", err)
+	}
+
+	log = log.WithValues("controlPlane", controlPlaneRef.Name)
+
+	patchHelper, err := patch.NewHelper(awsManagedCluster, r.Client)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to init patch helper: %w", err)
+	}
+
+	// Set the values from the managed control plane
+	awsManagedCluster.Status.Ready = controlPlane.Status.Ready
+	awsManagedCluster.Spec.ControlPlaneEndpoint = controlPlane.Spec.ControlPlaneEndpoint
+	awsManagedCluster.Status.FailureDomains = controlPlane.Status.FailureDomains
+
+	if err := patchHelper.Patch(ctx, awsManagedCluster); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to patch AWSManagedCluster: %w", err)
+	}
+
+	log.Info("Successfully reconciled AWSManagedCluster")
+
+	return reconcile.Result{}, nil
+
+}
+
+func (r *AWSManagedClusterReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	awsManagedCluster := &infrav1exp.AWSManagedCluster{}
+
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(awsManagedCluster).
+		WithEventFilter(predicates.ResourceNotPaused(r.Log)).
+		Build(r)
+
+	if err != nil {
+		return fmt.Errorf("error creating controller: %w", err)
+	}
+
+	// Add a watch for clusterv1.Cluster unpaise
+	if err = controller.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: util.ClusterToInfrastructureMapFunc(awsManagedCluster.GroupVersionKind()),
+		},
+		predicates.ClusterUnpaused(r.Log),
+	); err != nil {
+		return fmt.Errorf("failed adding a watch for ready clusters: %w", err)
+	}
+
+	// Add a watch for AWSManagedControlPlane
+	if err = controller.Watch(
+		&source.Kind{Type: &infrav1exp.AWSManagedControlPlane{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(r.managedControlPlaneToManagedCluster),
+		},
+		predicates.ClusterUnpaused(r.Log),
```

Comment:
This predicate will try to cast the object into a `Cluster` and throw an error:

```
[manager] I0824 16:03:56.866549       7 cluster_predicates.go:131] controllers/AWSManagedCluster "msg"="Expected Cluster" "eventType"="update" "predicate"="ClusterUpdateUnpaused" "type"="infrastructure.cluster.x-k8s.io/v1alpha3, Kind=AWSManagedControlPlane"
```

I think this is why we're seeing AWSManagedCluster not reconcile immediately when AWSManagedControlPlane status is set to Ready.

Reply:
Thats it, thanks.

Reply:
Removed the ClusterUnpaused